kliff/models/parameter.py

```diff
@@ -1,705 +1,390 @@
+                bounds are applied
+            is_mutable: boolean if the parameter is mutable. This implies that either
+                the parameter will be passed to the scipy optimizer, or the parameter values will be edited.
+            index: Index of the parameter in the parameter vector. Used for setting the
+             parameter in the KIMPY.
+            opt_mask: Boolean array of the same shape as the parameter. The values
+                marked ``True`` are optimized, and ``False`` are not optimized.
 
-    def set_lower_bound(self, index: int, v: float):
-        """
-        Set the lower bound of a component of the parameter.
+        Returns:
+            A new instance of Parameter.
+        """
+        array_in = deepcopy(np.asarray(input_array))
+        obj = array_in.view(cls)
+        obj.name = name
+        obj.transform = transform
+        obj.index = index
+        obj._is_transformed = False
+        obj.bounds = bounds
+        obj.is_mutable = is_mutable
+        if opt_mask is not None:
+            obj.opt_mask = opt_mask
+        else:
+            obj.opt_mask = np.ones(obj.shape, dtype=bool)
+        obj._bounds_transformed = False
+        return obj
+
+    # TODO: This seems a bit off, the signature should match np.array, need to look more in it
+    # but the format matches the one in numpy examples.
+    def __array_finalize__(self, obj):
+        """Finalizes a parameter, needed for numpy object cleanup."""
+        if obj is None:
+            return
+        self.name = getattr(obj, "name", None)
+        self.transform = getattr(obj, "transform", None)
+        self.bounds = getattr(obj, "bounds", None)
+        self.is_mutable = getattr(obj, "is_mutable", False)
+        self.index = getattr(obj, "index", None)
+        self._is_transformed = getattr(obj, "_is_transformed", False)
+        self.opt_mask = getattr(obj, "opt_mask", None)
+        self._bounds_transformed = getattr(obj, "_bounds_transformed", False)
+
+    def __repr__(self):
+        return f"Parameter {self.name} {np.ndarray.__repr__(self)}."
+
+    def transform_(self):
+        """Apply the transform to the parameter.
+
+        This method simple applies the function ~:kliff.transforms.ParameterTransform.__call__
+        to the parameter (or equivalently, ~:kliff.transforms.ParameterTransform.transform).
+        Difference between :transform_ and :transform is that :transform_ does inplace
+        transformation, while :transform returns a new transformed parameter.
+        """
+        if self._is_transformed:
+            # warnings.warn("Parameter {0} has already been transformed.".format(self.name))
+            # Warnings become quite noisy, so commenting it out for now.
+            # TODO: figure out a better solution for this.
+            return
+        else:
+            if self.transform is not None:
+                for i in range(len(self)):
```

Comment:
The major reason to loop was to inplace modify the values. For example, an expression such as
```
self = func(self)
```
Will compute new values and overwrite the self pointers. It was suggested online that this can result in situations where your array is implicitly converted to a plain array etc. But I have not tested the validity of that argument extensively, and decided to play safe.

I dont have any preference either way so, can try full array functions once. If it does not work, will revert. Leaving this comment unresolved for now to remind.

Reply:
Just saw that it was not online forums, but warning my IDE!
![image](https://github.com/openkim/kliff/assets/6790415/c2e333a6-6c8e-4910-9c2a-3cf133752f46)
I think better to do a deep copy of it and assign element wise.

Reply:
Just to add it to the discussion, another reason for doing elementwise transformations was to avoid situations where the transformation might be a R^n to R^m function. But I guess thats upto user to provide correctly.